src/presence/heartbeat.js

```diff
@@ -0,0 +1,61 @@
+// @flow strict-local
+
+/**
+ * Heartbeat: Perform callbacks at regular intervals.
+ *
+ * While active, calls `callback` with `true` every `milliseconds` milliseconds.
+ * While inactive, does nothing.
+ *
+ * When transitioning between active and inactive states, performs one
+ * additional edge-triggered callback with `true` (if starting) or `false` (if
+ * stopping).
```

Comment:
Per the discussion linked and quoted above, I think we want one of these and not the other.

That probably makes most sense as a separate later commit anyway, though -- trying to make the switch to this implementation as close to NFC as it reasonably can be.

Reply:
Ironically, the current observed (buggy) behavior is closer to the quote above than this implementation of Heartbeat is. It would be more `[nfc]`, from a certain perspective, to go straight there.

(I'd still rather fix things _before_ changing them, though.)

Reply:
> Ironically, the current observed (buggy) behavior is closer to the quote above than this implementation of Heartbeat is. It would be more `[nfc]`, from a certain perspective, to go straight there.

Hmm, I guess that's right! That might be a clean way to do it too, then.


> (I'd still rather fix things *before* changing them, though.)

Not sure I understand; isn't a fix a change?

Reply:
> Not sure I understand; isn't a fix a change?

A change to the code, but not necessarily to its specification. Perhaps read as: I'd rather fix the existing code (leaving the intended specification unchanged) before separately changing that specification (and the code along with it).

(This is decidedly situational. If, for example, the second step would have involved eliminating all the code in the first step, I would probably have a very different opinion.)

Reply:
Got it. If I now understand you right, I think that means that

> (I'd still rather fix things _before_ changing them, though.)

was agreeing with

> That probably makes most sense as a separate later commit anyway, though

Is that right? I didn't read it that way the first time. :slightly_smiling_face: